src/Paramore.Brighter/MessageMappers/JsonMessageMapper.cs

```diff
@@ -0,0 +1,47 @@
+using System;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace Paramore.Brighter.MessageMappers;
+
+public class JsonMessageMapper<TRequest>(RequestContext? context) : IAmAMessageMapper<TRequest>, IAmAMessageMapperAsync<TRequest> where TRequest : class, IRequest
+{
+    public IRequestContext? Context { get; set; } = context;
+
+    public Task<Message> MapToMessageAsync(TRequest request, Publication publication,
+        CancellationToken cancellationToken = default)
+        => Task.FromResult(MapToMessage(request, publication));
+
+    public Task<TRequest> MapToRequestAsync(Message message, CancellationToken cancellationToken = default)
+        => Task.FromResult(MapToRequest(message));
+
+    public Message MapToMessage(TRequest request, Publication publication)
+    {
+        MessageType messageType = request switch
+        {
+            ICommand => MessageType.MT_COMMAND,
+            IEvent => MessageType.MT_EVENT,
+            _ => throw new ArgumentException(@"This message mapper can only map Commands and Events", nameof(request))
+        };
+
+        if(publication.Topic is null)
+            throw new ArgumentException($"No Topic Defined for {publication}");
+
+        var header = new MessageHeader(messageId: request.Id, topic: publication.Topic, messageType: messageType);
+
+        var body = new MessageBody(JsonSerializer.Serialize(request, JsonSerialisationOptions.Options));
```

Comment:
In principle there is an async serializer. I am not sure if it actually has meaning. But we should investigate. The MS route here is to pass an enum flag to the common method and then just use sync or async code in the body, based on that flag as needed

Reply:
See the Flag Argument Hack here: https://learn.microsoft.com/en-us/archive/msdn-magazine/2015/july/async-programming-brownfield-async-development